### PR TITLE
feat(core): Add cache to dgraph.type predicate.

### PR DESCRIPTION
**Pull request by harshil-goel**
Add cache to dgraph.type predicate. This would significantly help improve @filter(type) queries. 
Live loader before took: 16mins 39 seconds for 21 million dataset.
Live loader now: 10mins 52 seconds.
Improvement: 38%

### CODE REVIEW

File: posting/mvcc.go [hunk 1]
```diff
@@ -457,33 +492,62 @@ func ReadPostingList(key []byte, it *badger.Iterator) (*List, error) {
...
+
+	pk, _ := x.Parse(key)
+	countMap.Lock()
+	countMap.count[string(key)]++
+
+	// We use badger subscription to invalidate the cache. For every write we make the value
+	// corresponding to the key in the cache to nil. So, if we get some non-nil value from the cache
+	// then it means that no  writes have happened after the last set of this key in the cache.
+	if l, ok := countMap.list[string(key)]; ok {
+		if l != nil && l.maxTs < readTs {
```
Comment (harshil-goel) [inline posting/mvcc.go:524]: l.minTs > readTs

File: posting/list.go [hunk 2]
```diff
@@ -762,6 +791,38 @@ func (l *List) IsEmpty(readTs, afterUid uint64) (bool, error) {
 	return count == 0, nil
 }
 
+func (l *List) getPostingAndLengthNoSort(readTs, afterUid, uid uint64) (int, bool, *pb.Posting) {
```
Comment (mangalaman93) [inline posting/list.go:794]: we should create a separate PR for this

File: posting/mvcc.go [hunk 3]
```diff
@@ -1,25 +1,24 @@
 /*
```
Comment (mangalaman93) [inline posting/mvcc.go:1]: undo this change

File: posting/mvcc.go [hunk 4]
```diff
@@ -59,6 +58,14 @@ type incrRollupi struct {
 	closer   *z.Closer
 }
 
+type GlobalCache struct {
+	sync.RWMutex
+
+	count      map[string]int
```
Comment (mangalaman93) [inline posting/mvcc.go:64]: use a struct instead! More maps => more memory unpredictability

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
See whether we want to put this behind a feature flag just in case!

**Review by mangalaman93 — APPROVED** [review]

**Pull request merged**